etk-asm/src/asm.rs

```
@@ -244,107 +231,127 @@ impl Assembler {
-                        .concretize((&self.declared_labels, &self.declared_macros).into())
-                    {
-                        Ok(_) => unreachable!(),
-                        Err(ops::Error::ContextIncomplete {
-                            source: expression::Error::UnknownMacro { name, .. },
-                            ..
-                        }) => error::UndeclaredExpressionMacro { name }.fail(),
-                        Err(ops::Error::ContextIncomplete {
-                            source: expression::Error::UnknownLabel { .. },
-                            ..
-                        }) => {
-                            let labels = op.expr().unwrap().labels(&self.declared_macros).unwrap();
-                            let declared = self.declared_labels.into_keys().collect();
-                            let invoked: HashSet<_> = labels.into_iter().collect();
-                            let missing = invoked
-                                .difference(&declared)
-                                .cloned()
-                                .collect::<Vec<String>>();
-                            error::UndeclaredLabels { labels: missing }.fail()
+    /// Inspect the macros in a series of [`RawOp`] and declare them in the assembler.
+    fn inspect_macros<I, O>(&mut self, nodes: &I) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = O> + Clone,
+        O: Into<RawOp>,
+    {
+        for op in nodes.clone() {
+            let op = op.into();
+            if let RawOp::Op(AbstractOp::MacroDefinition(_)) = op {
+                self.declare_macro(op)?
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Collect any assembled instructions that are ready to be output.
+    fn take(&mut self) -> Vec<u8> {
+        let output = self.concretize_ops();
+        match output {
+            Ok(v) => {
+                self.ready.clear();
+                v
+            }
+            Err(_) => Vec::new(),
```

This looks like it discards errors from `concretize_ops`.